trie/triedb/pathdb/history.go

```diff
@@ -0,0 +1,594 @@
+func (h *history) decode(accountData, storageData, accountIndexes, storageIndexes []byte) error {
+	var (
+		accounts    = make(map[common.Hash][]byte)
+		storages    = make(map[common.Hash]map[common.Hash][]byte)
+		accountList []common.Hash
+		storageList = make(map[common.Hash][]common.Hash)
+
+		r = &decoder{
+			accountData:    accountData,
+			storageData:    storageData,
+			accountIndexes: accountIndexes,
+			storageIndexes: storageIndexes,
+		}
+	)
+	if err := r.verify(); err != nil {
+		return err
+	}
+	for i := 0; i < len(accountIndexes)/accountIndexSize; i++ {
+		// Resolve account first
+		aIndex, aData, err := r.readAccount(i)
+		if err != nil {
+			return err
+		}
+		accounts[aIndex.Hash] = aData
+		accountList = append(accountList, aIndex.Hash)
+
+		// Resolve storage slots
+		slotList, slotData, err := r.readStorage(aIndex)
+		if err != nil {
+			return err
+		}
+		if len(slotList) > 0 {
+			storageList[aIndex.Hash] = slotList
+			storages[aIndex.Hash] = slotData
+		}
+	}
+	h.accounts = accounts
+	h.accountList = accountList
+	h.storages = storages
+	h.storageList = storageList
+	return nil
+}
+
+// readHistory reads and decodes the state history object by the given id.
+func readHistory(freezer *rawdb.ResettableFreezer, id uint64) (*history, error) {
+	blob := rawdb.ReadStateHistoryMeta(freezer, id)
+	if len(blob) == 0 {
+		return nil, fmt.Errorf("state history not found %d", id)
+	}
+	var m meta
+	if err := m.decode(blob); err != nil {
+		return nil, err
+	}
+	var (
+		dec            = history{meta: &m}
+		accountData    = rawdb.ReadStateAccountHistory(freezer, id)
+		storageData    = rawdb.ReadStateStorageHistory(freezer, id)
+		accountIndexes = rawdb.ReadStateAccountIndex(freezer, id)
+		storageIndexes = rawdb.ReadStateStorageIndex(freezer, id)
+	)
+	if err := dec.decode(accountData, storageData, accountIndexes, storageIndexes); err != nil {
+		return nil, err
+	}
+	return &dec, nil
+}
+
+// writeStateHistory writes the state history with provided state set. After
+// storing the corresponding state history, it will also prune the stale
+// histories from the disk with the given threshold.
+func writeStateHistory(freezer *rawdb.ResettableFreezer, dl *diffLayer, limit uint64) error {
+	// Short circuit if state set is not available.
+	if dl.states == nil {
+		return errors.New("state change set is not available")
+	}
+	var (
+		err   error
+		n     int
+		start = time.Now()
+		h     = newHistory(dl.rootHash, dl.parent().root(), dl.blockNumber, dl.states)
+	)
+	accountData, storageData, accountIndex, storageIndex := h.encode()
+	dataSize := common.StorageSize(len(accountData) + len(storageData))
+	indexSize := common.StorageSize(len(accountIndex) + len(storageIndex))
+
+	// Write history data into five freezer table respectively.
+	rawdb.WriteStateHistory(freezer, dl.stateID(), h.meta.encode(), accountIndex, storageIndex, accountData, storageData)
```

Comment:
Any reason why writing the state history is one operation, but reading it is 5 separate operations on rawdb? Do we ever need only part of a state history? If not, can't we just pull is all up in one go?

Reply:
History metadata will be loaded separately when we want to determine if a specific state is recoverable.
Usually we will load a batch of meta data in a single freezer read. So I will keep the read API for meta.


For other parts(account index, account data, storage index, storage data), they will also be loaded on demand
when we want to serve an archive state request.

For example, if I want to retrieve account X in the history entry W, what I need is
- the entire account index belonging to W
- partial account data belonging to W(located by offset and length from account index)

With these standalone read APIs, I don't need to load the unused data which is expensive.

Reply:
👍